examples/capture/main.rs

```diff
@@ -0,0 +1,98 @@
+    });
+
+    let mut device = adapter.request_device(&wgpu::DeviceDescriptor {
+        extensions: wgpu::Extensions {
+            anisotropic_filtering: false,
+        },
+        limits: wgpu::Limits::default(),
+    });
+
+    // Rendered image is 256×256 with 32-bit BGRA color
+    let size = 256u32;
+
+    // The output buffer lets us retrieve the data as an array
+    let output_buffer = device.create_buffer(&wgpu::BufferDescriptor {
+        size: (size * size * std::mem::size_of::<u32>() as u32) as u64,
+        usage: wgpu::BufferUsage::MAP_READ,
+    });
+
+    let texture_extent = wgpu::Extent3d {
+        width: size,
+        height: size,
+        depth: 1,
+    };
+
+    // The render pipeline renders data into this texture
+    let texture = device.create_texture(&wgpu::TextureDescriptor {
+        size: texture_extent,
+        array_layer_count: 1,
+        mip_level_count: 1,
+        sample_count: 1,
+        dimension: wgpu::TextureDimension::D2,
+        format: wgpu::TextureFormat::Bgra8Unorm,
+        usage: wgpu::TextureUsage::NONE,
+    });
+
+    // Set the background to be red
+    let command_buffer = {
+        let mut encoder =
+            device.create_command_encoder(&wgpu::CommandEncoderDescriptor { todo: 0 });
+        encoder.begin_render_pass(&wgpu::RenderPassDescriptor {
+            color_attachments: &[wgpu::RenderPassColorAttachmentDescriptor {
+                attachment: &texture.create_default_view(),
+                resolve_target: None,
+                load_op: wgpu::LoadOp::Clear,
+                store_op: wgpu::StoreOp::Store,
+                clear_color: wgpu::Color::RED,
+            }],
+            depth_stencil_attachment: None,
+        });
+
+        // Copy the data from the texture to the buffer
+        encoder.copy_texture_to_buffer(
+            wgpu::TextureCopyView {
+                texture: &texture,
+                mip_level: 0,
+                array_layer: 0,
+                origin: wgpu::Origin3d::ZERO,
+            },
+            wgpu::BufferCopyView {
+                buffer: &output_buffer,
+                offset: 0,
+                row_pitch: std::mem::size_of::<u32>() as u32 * size,
+                image_height: size,
+            },
+            texture_extent,
+        );
+
+        encoder
+    }
+    .finish();
+
+    device.get_queue().submit(&[command_buffer]);
+
+    // Loop over each pixel in the buffer, checking that it is red
+    output_buffer.map_read_async(
+        0,
+        (std::mem::size_of::<u32>() as u32 * size * size) as u64,
+        |result: wgpu::BufferMapAsyncResult<&[u8]>| {
+            for pixel in result.unwrap().data.chunks(std::mem::size_of::<u32>()) {
+                let expected_b = 0;
+                let expected_g = 0;
+                let expected_r = 255;
+                let expected_a = 255;
+                assert_eq!(pixel, [expected_b, expected_g, expected_r, expected_a]);
+            }
+        },
```

Comment:
I think we need a `device.poll(true);` here to ensure the map_read_async runs.

Reply:
Okay, I added this, hopefully in the correct place.

Reply:
@rukai we wait for all the callbacks on device destruction, so this shouldn't be required. But I think it's still useful to `poll` explicitly here for demonstration only.

Reply:
Oh! Must be a bug then, because I have always had to include this to ensure the map_read_async runs. Created https://github.com/gfx-rs/wgpu-rs/issues/41